### PR TITLE
Merge bodies on renaming conflicts

### PR DESCRIPTION
**Pull request by gaabs**
This aims to address #52 by also detecting conflicts when "one developer renames a declaration and changes its body, and the other developer changes the body of the same declaration, in the same area"

Note that it also counts each conflict on body as a Renaming Conflict, potentially increasing the number of reported conflicts.

### CODE REVIEW

File: src/main/java/br/ufpe/cin/mergers/util/RenamingUtils.java [hunk 1]
```diff
@@ -93,30 +97,67 @@ public static boolean haveSameParent(FSTNode left, FSTNode right) {
...
+    public static String mergeBodies(String leftContent, String baseContent, String rightContent) {
+        String bodyResult;
+
+        try {
+            String leftBody = RenamingUtils.removeSignature(leftContent);
+            String baseBody = RenamingUtils.removeSignature(baseContent);
+            String rightBody = RenamingUtils.removeSignature(rightContent);
+            bodyResult = TextualMerge.merge(leftBody, baseBody, rightBody, JFSTMerge.isWhitespaceIgnored);
+        } catch (TextualMergeException e) {
+            e.printStackTrace();
```
Comment (pauloborba) [inline src/main/java/br/ufpe/cin/mergers/util/RenamingUtils.java]: @gaabs segue-se em frente mesmo tendo dado problema no textual?

File: src/main/java/br/ufpe/cin/mergers/util/RenamingUtils.java [hunk 2]
```diff
@@ -93,30 +97,67 @@ public static boolean haveSameParent(FSTNode left, FSTNode right) {
...
-            context.deletionConflicts++;
-        } else {
-            context.renamingConflicts++;
+        //first creates a conflict
+        String newConflictBody;
+        try {
+            newConflictBody = TextualMerge.merge(firstContent, baseContent, secondContent, JFSTMerge.isWhitespaceIgnored);
+        } catch (TextualMergeException e) {
+            e.printStackTrace();
+            MergeConflict newConflict = new MergeConflict(firstContent + '\n', secondContent + '\n');
```
Comment (pauloborba) [inline src/main/java/br/ufpe/cin/mergers/util/RenamingUtils.java]: @gaabs segue-se em frente mesmo tendo dado problema no textual?

**Comment by guilhermejccavalcanti** [comment]
@gaabs  where in the code is this actually addressed _"one developer renames a declaration and changes its body, and the other developer changes the body of the same declaration, in the same area"_ ? I could not figure it out, and what is the reasoning behind the solution?

I am missing comments explaining some decisions. The handlers are not trivial, not even for us responsible for them, so I think comments "here, and there" and in tricky places could help.

**Comment by gaabs** [comment]
> @gaabs where in the code is this actually addressed _"one developer renames a declaration and changes its body, and the other developer changes the body of the same declaration, in the same area"_ ? I could not figure it out, and what is the reasoning behind the solution?
> 
> I am missing comments explaining some decisions. The handlers are not trivial, not even for us responsible for them, so I think comments "here, and there" and in tricky places could help.

You are right, @guilhermejccavalcanti , the problem has not been fully addressed.

I will describe the present behaviour, both for SAFE and MergeMethods handlers, on single renamings, and then the proposed changes.

The Safe Renaming Handler (the default s3m renaming handler), only reports conflict when unstructured merge reports a conflict that include the whole method signature. When that does not happen, the handler keeps both versions of the method, even if there was only a conflict on the method's body. I believe this was one of the points of the issue, which has not been addressed yet. For this case, it would be necessary to know, for each unstructured conflict, which method it refers to, so that a proper match on handlers could be made.

Also, the conflict reported by the safe handler include the entire versions of each side, even if only minor conflicts happen on the body. This PR fixes this behavior, by applying the textual merge.

For the MergeMethods handler, the old behavior was to merge the modified signature with the body from the other side. The problem was that it did not consider conflicts on method bodies. This PR fixes that by applying TextualMerge on bodies of both versions.

**Comment by jvcoutinho** [comment]
@guilhermejccavalcanti @pauloborba Summary of this PR:

- The problem is solved in the MergeMethods version of the renaming handler. It calls textual merge inside the body of the methods.
- The problem is not solved in the Safe (default) version of the renaming handler, because it reports a conflict only if textual merge reports a conflict that includes the method signature (instead of only the body). Then, it keeps both methods.

**Comment by pauloborba** [comment]
@jvcoutinho this has been superseded by the PR you submitted, right?

**Comment by jvcoutinho** [comment]
> @jvcoutinho this has been superseded by the PR you submitted, right?
Yes, @pauloborba.

**Comment by pauloborba** [comment]
Supersed by another PR

**Pull request closed without merging**